aten/src/ATen/native/cpu/radix_sort.h

```diff
@@ -0,0 +1,196 @@
+    local_histogram[(key_2 >> (pass * 8)) & 0xFF]++;
+    local_histogram[(key_3 >> (pass * 8)) & 0xFF]++;
+    local_histogram[(key_4 >> (pass * 8)) & 0xFF]++;
+  }
+  if (tid == (nthreads - 1)) {
+    for (int64_t i = elements_count_4; i < elements_count; ++i) {
+      K key = input_keys[i];
+      local_histogram[(key >> (pass * 8)) & 0xFF]++;
+    }
+  }
+#pragma omp barrier
+  // Step 2: prefix sum
+  if (tid == 0) {
+    int sum = 0, prev_sum = 0;
+    for (int bins = 0; bins < RDX_HIST_SIZE; bins++) {
+      for (int t = 0; t < nthreads; t++) {
+        sum += histogram[t * RDX_HIST_SIZE + bins];
+        histogram_ps[t * RDX_HIST_SIZE + bins] = prev_sum;
+        prev_sum = sum;
+      }
+    }
+    histogram_ps[RDX_HIST_SIZE * nthreads] = prev_sum;
+    TORCH_CHECK(prev_sum == elements_count);
+  }
+#pragma omp barrier
+
+  // Step 3: scatter
+#pragma omp for schedule(static)
+  for (int64_t i = 0; i < elements_count_4; i += 4) {
+    K key_1 = input_keys[i];
+    K key_2 = input_keys[i + 1];
+    K key_3 = input_keys[i + 2];
+    K key_4 = input_keys[i + 3];
+
+    int bin_1 = (key_1 >> (pass * 8)) & 0xFF;
+    int bin_2 = (key_2 >> (pass * 8)) & 0xFF;
+    int bin_3 = (key_3 >> (pass * 8)) & 0xFF;
+    int bin_4 = (key_4 >> (pass * 8)) & 0xFF;
+
+    int pos;
+    pos = local_histogram_ps[bin_1]++;
+    output_keys[pos] = key_1;
+    output_values[pos] = input_values[i];
+    pos = local_histogram_ps[bin_2]++;
+    output_keys[pos] = key_2;
+    output_values[pos] = input_values[i + 1];
+    pos = local_histogram_ps[bin_3]++;
+    output_keys[pos] = key_3;
+    output_values[pos] = input_values[i + 2];
+    pos = local_histogram_ps[bin_4]++;
+    output_keys[pos] = key_4;
+    output_values[pos] = input_values[i + 3];
+  }
+  if (tid == (nthreads - 1)) {
+    for (int64_t i = elements_count_4; i < elements_count; ++i) {
+      K key = input_keys[i];
+      int pos = local_histogram_ps[(key >> (pass * 8)) & 0xFF]++;
+      output_keys[pos] = key;
+      output_values[pos] = input_values[i];
+    }
+  }
+}
+
+} // namespace
+
+constexpr bool is_radix_sort_available() { return true; }
+
+template <typename K, typename V>
+std::pair<K*, V*> radix_sort_parallel(
+    K* inp_key_buf,
+    V* inp_value_buf,
+    K* tmp_key_buf,
+    V* tmp_value_buf,
+    int64_t elements_count,
+    int64_t max_value) {
+  int maxthreads = omp_get_max_threads();
+  std::unique_ptr<int []> histogram_tmp(new int[RDX_HIST_SIZE * maxthreads]);
+  std::unique_ptr<int []> histogram_ps_tmp(new int[RDX_HIST_SIZE * maxthreads + 1]);
+  int* histogram = histogram_tmp.get();
+  int* histogram_ps = histogram_ps_tmp.get();
+  if (max_value == 0) {
+    return std::make_pair(inp_key_buf, inp_value_buf);
+  }
+
+  // __builtin_clz is not portable
+  int num_bits = sizeof(K) * 8 - llvm::countLeadingZeros(static_cast<std::make_unsigned_t<K>>(max_value));
+  unsigned int num_passes = (num_bits + 7) / 8;
+
+#pragma omp parallel
```

Comment:
You should use `parallel_for` here, not use `#pragma omp` here, so that other thread pool implementations can be easily swapped out

Reply:
this algo requires sync between threads (with `omp barrier)`, the problem is `parallel_for` can't do this at the moment.

if i use `parallel_for`, i need to do sync outside the omp session, kind of like:
```
## this is going to have a lot of overhead with omp thread join
step 1: parallel_for (omp session 1)
step 2: sync outside the omp session
step 3: paralle_for (omp session 2)
step 4: sync outside the omp session
   ... ...
```

so if we use `omp barrier`, all the above job can be done with only one omp session.

This is the same story why we don't have paralleled kernel for `cumsum()` and `log_softmax()` when the input is a 1D tensor, they all require sync within the omp session to achieve good performance.

Reply:
Hmm, as a follow up, can we add the capability for syncing in `parallel_for`? Will have to figure out if this can actually be implemented in an agnostic way...

Reply:
cc @cpuhrsch

Reply:
> Hmm, as a follow up, can we add the capability for syncing in `parallel_for`? Will have to figure out if this can actually be implemented in an agnostic way...

Unfortunately no, barrier is not supported by thread runtimes like TBB or Eigen...